iree/compiler/Conversion/LinalgToSPIRV/VectorizeMemref.cpp

```diff
@@ -360,6 +364,38 @@ class ProcessPlaceHolder final
   }
 };
 
+// Folds linalg.reshape op that directly reshaping an iree.placeholder op into
+// the iree.placeholder op itself.
+class FoldReshapeIntoPlaceholder final
+    : public OpRewritePattern<linalg::ReshapeOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::ReshapeOp reshapeOp,
+                                PatternRewriter &rewriter) const override {
+    auto placeholderOp = reshapeOp.src().getDefiningOp<IREE::PlaceholderOp>();
+    if (!placeholderOp) return failure();
+    rewriter.replaceOpWithNewOp<IREE::PlaceholderOp>(
+        reshapeOp, reshapeOp.getResultType(), ValueRange(),
+        placeholderOp.getAttrs());
+    return success();
+  }
+};
+
+/// Removes iree.placeholder ops that has no uses.
+class RemoveDeadPlaceholder final
```

Comment:
This pass should do this automatically https://github.com/google/iree/blob/main/iree/compiler/Conversion/Common/RemoveDeadMemAllocsPass.cpp

Reply:
That's a pass though. Here the unused iree.placeholder is left after running by the above reshape folding pattern. So I need a pattern to clean it up before running next steps. So cannot rely on that pass. I can create a `populateRemoveDeadMemAllocsPatterns` entry point to pull that pattern, but not sure it's worth it given this pattern is just 4 LOC canonicalization.

Reply:
If you move the above pattern to HLOToLinalgOnBuffers, and run this pass before TileAndFuse pass that should work